### PR TITLE
EVE instances for different configs

### PR DESCRIPTION
**Pull request by sadov**
Added support for launching multiple instances of EVE for different configurations:
```
make build-tests
./eden config add default
./eden config add t1
./eden config set t1 --key eve.hostfwd --value '{"2223":"22"}'
./eden config set t1 --key eve.telnet-port --value 7778
./eden setup -v debug # default config
./eden setup --config t1 -v debug
./eden start -v debug # start all services and EVE with default config
./eden eve start --config t1 -v debug # start second EVE with t1 config
```

Signed-off-by: Oleg Sadov <oleg.sadov@gmail.com>

### CODE REVIEW

**Comment by mydatascience** [comment]
Ehm I see only one eve instance. Could you please explain how to make two?

**Comment by giggsoff** [comment]
> Ehm I see only one eve instance. Could you please explain how to make two?

As I can see, we have `default` and `t1` contexts. That is, two instances of EVE.

**Comment by sadov** [comment]
Actually you can log into both EVEs through the console interface and see different telnet ports (as we set for instances) and UUIDs in the prompts (the same as we can see in `eve onboard`). Like this:
```
$ ./eden eve console
INFO[0000] Try to telnet 127.0.0.1:7777                 
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.

002388f2-abbf-440a-b4b6-07c832587dd0:~# 
```
and
```
$ eden eve console --config t1 -v debug
configName:  t1
configFile:  /home/user/.eden/contexts/t1.yml
DEBU[0000] Will use config from /home/user/.eden/contexts/default.yml 
DEBU[0000] Try to add config from /home/user/work/EVE/github/itmo-eve/eden/eden-config.yml 
DEBU[0000] Will use config from /home/user/.eden/contexts/t1.yml 
DEBU[0000] Try to add config from /home/user/work/EVE/github/itmo-eve/eden/eden-config.yml 
INFO[0000] Try to telnet 127.0.0.1:7778                 
Trying 127.0.0.1...
Connected to 127.0.0.1.
Escape character is '^]'.

1b4fcc3c-eb02-4461-9ee5-30de8c06d993:~#
```

File: cmd/edenClean.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ var cleanCmd = &cobra.Command{
 			redisPort = viper.GetInt("redis.port")
 			redisDist = utils.ResolveAbsPath(viper.GetString("redis.dist"))
 			eveImageFile = utils.ResolveAbsPath(viper.GetString("eve.image-file"))
-			evePidFile = utils.ResolveAbsPath(viper.GetString("eve.pid"))
+			evePidFile = utils.ResolveAbsPath(configName + "-" + viper.GetString("eve.pid"))
```
Comment (giggsoff) [inline cmd/edenClean.go:43]: Probably, we need the same [here](https://github.com/lf-edge/eden/blob/22f78cc2c39181edc9ca69973fa67c2fc4c3311a/cmd/eve.go#L58) and [here](https://github.com/lf-edge/eden/blob/22f78cc2c39181edc9ca69973fa67c2fc4c3311a/cmd/eve.go#L147)?

Comment (giggsoff) [reply cmd/edenClean.go:43]: The [same](https://github.com/lf-edge/eden/blob/22f78cc2c39181edc9ca69973fa67c2fc4c3311a/cmd/eve.go#L217)

**Comment by mydatascience** [comment]
> > Ehm I see only one eve instance. Could you please explain how to make two?
> 
> As I can see, we have `default` and `t1` contexts. That is, two instances of EVE.

I don't see default

If we consider this implicitly please describe explicitly. I believe then we need to add 
`eden config add default `

Also I don't understand the meaning of  
```
./eden start -v debug
./eden start --config t1 -v debug
```

what should it do?

File: cmd/edenClean.go (hunk 1 above)
Comment (sadov) [reply cmd/edenClean.go:43]: Probably - it needs to be seen.

**Comment by sadov** [comment]
> > > Ehm I see only one eve instance. Could you please explain how to make two?
> > 
> > 
> > As I can see, we have `default` and `t1` contexts. That is, two instances of EVE.
> 
> I don't see default
> 
> If we consider this implicitly please describe explicitly. I believe then we need to add
> `eden config add default `

Actually, we already have it after `make build-tests`.
 
> Also I don't understand the meaning of
> 
> ```
> ./eden start -v debug
> ./eden start --config t1 -v debug
> ```
> 
> what should it do?

Just starting all EDEN-related dockers and both EVE QEMUs instances.

File: cmd/edenClean.go (hunk 1 above)
Comment (sadov) [reply cmd/edenClean.go:43]: Fixed

Comment (sadov) [reply cmd/edenClean.go:43]: Seems -- some Viper's black magic:
```
$ ./eden start -v debug
...
FATA[0000] open /home/user/work/EVE/github/itmo-eve/eden/dist/default-/home/user/work/EVE/github/itmo-eve/eden/dist/default-eve.log: no such file or directory 
ERRO[0014] cannot start eve: exit status 1
```
Need to look...

**Comment by mydatascience** [comment]
Please change the instruction to 
1)  Have explicitly 2 configs 
2) Have 1 start. And second start ./eden eve start --config t1

**Comment by sadov** [comment]
Done

**Comment by giggsoff** [comment]
They are started and onboarded, great!

Bu, unfortunately, `process doesn't exist` for the second one:
```
giggsoff@giggsoff-PC:~/itmoeve/githubeden$ ./eden eve status
✔ EVE REMOTE IPs: 192.168.0.10; fec0::466:7cd4:7122:7db9; fe80::5d7b:b2d9:1032:8b56
        Last info received time: 2020-12-03 15:58:17 +0300 MSK
✔ EVE on Qemu status: running with pid 230176
        Logs for local EVE at: /home/giggsoff/itmoeve/githubeden/dist/default-eve.log
✔ EVE Request IP: 192.168.31.137
giggsoff@giggsoff-PC:~/itmoeve/githubeden$ ./eden eve status --config=t1
✔ EVE REMOTE IPs: 192.168.0.10; fec0::7ba5:602e:8499:359b; fe80::e7cc:2cc8:70da:b6f3
        Last info received time: 2020-12-03 15:58:25 +0300 MSK
✘ EVE on Qemu status: process doesn't exist
        Logs for local EVE at: /home/giggsoff/itmoeve/githubeden/dist/t1-eve.log
✔ EVE Request IP: 192.168.31.137
```
And I cannot see files:
```
giggsoff@giggsoff-PC:~/itmoeve/githubeden$ ls dist/*.pid
dist/default-eve.pid  dist/eve.pid
giggsoff@giggsoff-PC:~/itmoeve/githubeden$ ls dist/*.log
dist/default-eve.log  dist/eve.log
```

File: pkg/eden/eden.go [hunk 2]
```diff
@@ -331,9 +331,9 @@ func StatusEServer() (status string, err error) {
 }
 
 //StartEVEQemu function run EVE in qemu
-func StartEVEQemu(commandPath string, qemuARCH string, qemuOS string, eveImageFile string, qemuSMBIOSSerial string, qemuAccel bool, qemuConfigFilestring, logFile string, pidFile string) (err error) {
-	commandArgsString := fmt.Sprintf("eve start --qemu-config=%s --eve-serial=%s --eve-accel=%t --eve-arch=%s --eve-os=%s --eve-log=%s --eve-pid=%s --image-file=%s -v %s",
-		qemuConfigFilestring, qemuSMBIOSSerial, qemuAccel, qemuARCH, qemuOS, logFile, pidFile, eveImageFile, log.GetLevel())
+func StartEVEQemu(commandPath string, configName string, qemuARCH string, qemuOS string, eveImageFile string, qemuSMBIOSSerial string, qemuAccel bool, qemuConfigFilestring, logFile string, pidFile string) (err error) {
+	commandArgsString := fmt.Sprintf("eve start --qemu-config=%s --eve-serial=%s --eve-accel=%t --eve-arch=%s --eve-os=%s --eve-log=%s --eve-pid=%s --image-file=%s -v %s --config %s",
+		qemuConfigFilestring, qemuSMBIOSSerial, qemuAccel, qemuARCH, qemuOS, logFile, pidFile, eveImageFile, log.GetLevel(), configName)
```
Comment (giggsoff) [inline pkg/eden/eden.go:336]: There is wrong sequence here.

**Review by giggsoff — CHANGES_REQUESTED** [review]
Please, see my comment and also modify https://github.com/lf-edge/eden/blob/22f78cc2c39181edc9ca69973fa67c2fc4c3311a/cmd/eve.go#L58-L59

**Comment by sadov** [comment]
> Please, see my comment and also modify
> 
> https://github.com/lf-edge/eden/blob/22f78cc2c39181edc9ca69973fa67c2fc4c3311a/cmd/eve.go#L58-L59

With a such modifications we have double adding of abs.path part as I reported at
https://github.com/lf-edge/eden/pull/407#discussion_r535036863

**Comment by giggsoff** [comment]
> > Please, see my comment and also modify
> > https://github.com/lf-edge/eden/blob/22f78cc2c39181edc9ca69973fa67c2fc4c3311a/cmd/eve.go#L58-L59
> 
> With a such modifications we have double adding of abs.path part as I reported at
> [#407 (comment)](https://github.com/lf-edge/eden/pull/407#discussion_r535036863)

So, we need resolve it, not leave as is.

**Comment by sadov** [comment]
> > > Please, see my comment and also modify
> > > https://github.com/lf-edge/eden/blob/22f78cc2c39181edc9ca69973fa67c2fc4c3311a/cmd/eve.go#L58-L59
> > 
> > 
> > With a such modifications we have double adding of abs.path part as I reported at
> > [#407 (comment)](https://github.com/lf-edge/eden/pull/407#discussion_r535036863)
> 
> So, we need resolve it, not leave as is.

It's reasonable

**Comment by giggsoff** [comment]
@sadov please take a look

**Comment by mydatascience** [comment]
I wanna merge it before the release. May I ?

**Comment by sadov** [comment]
> I wanna merge it before the release. May I ?

I'm looking now. On one test I got problems with onboarding for second config. I'm checking now on another comp.

**Comment by sadov** [comment]
That works for me.

**Pull request merged**